spock-core/src/main/java/org/spockframework/compiler/DeepBlockRewriter.java

```diff
@@ -145,11 +148,39 @@ private boolean handleImplicitCondition(ExpressionStatement stat) {
     checkIsValidImplicitCondition(stat);
     conditionFound = true;
     groupConditionFound = currSpecialMethodCall.isGroupConditionBlock();
+
+    if (currSpecialMethodCall.isWithCall()
+      && AstUtil.isInvocationWithImplicitThis(stat.getExpression())) {
+      replaceObjectExpressionWithCurrentClosure(stat);
+    }
+
     Statement condition = ConditionRewriter.rewriteImplicitCondition(stat, resources);
     replaceVisitedStatementWith(condition);
     return true;
   }
 
+  private void replaceObjectExpressionWithCurrentClosure(ExpressionStatement stat) {
+    MethodCallExpression methodCall = AstUtil.getExpression(stat, MethodCallExpression.class);
+    if (methodCall == null) return;
+
+    MethodCallExpression target = referenceToCurrentClosure();
+    methodCall.setObjectExpression(target);
+  }
+
+  @NotNull
```

Comment:
In that case I think it's better to remove the annotation since it's not used anywhere. 
Frankly, I confused it with "org.spockframework.util.Nullable" that is used in the same class DeepBlockRewriter::getCurrentWithOrMockClosure. I didn't check that NotNull and Nullable are derived from different libraries.